### PR TITLE
Account Creation Flow

### LINKED ISSUE #313: Account Creation
# Scope

Update initial user flow for selecting major and curriculum ([see figma](https://www.figma.com/file/sF9JJGhZDVWRs7lA9gBJTQ/Design?type=design&node-id=218-2102&mode=design&t=FQ34UABakDAFb8Rh-4)).

## TODO

- [x] Design Figma mockup
- [x] Get feedback (approve figma design)
- [x] Implement

### PR DESCRIPTION
**Pull request by zacktraczyk**
# Scope

User's need an intuitive way to select majors and minors when creating an account.

Closes #313.

# Implementation

- Created new login page at `/`
- Created new account creation flow at `/register`, `/register/majors`, and `/register/minors`

#### Important Notes:

- Used mix of routing redirect strategies
  - middleware handles authentication routing (i.e. `/` <-> `/register` / `planner`)
  - layout.tsx in register / planner routes handles routing based on `session.user.isRecordCreated` (i.e. `/register` <-> `/planner`)

- Not implementing help blurb in this PR
  - I realized this would require some work to get it to function how I want it, so going to split it into a separate PR

- Not implementing curriculum selection page implemented in this PR
  - I also realized updating your major or selecting a default curriculum now falls into the domain of other features with a different scope, so also going to make separate PRs for these

# How to test

#### Reset Account
- Logout
- Delete user record (`DELETE FROM "User" where "id"='< your user id >';`

#### Test Skip Account Creation
- (Reset Account)
- Click `continue with google`
- Select second option `skip account creation`
  - click continue (creates account w no majors / minors and reroutes to `/planner`)

#### Test Select Majors and Minors
- (Reset Account)
- Click `continue with google`
- Select first option `enter majors and minors`
  - click continue
- enter some majors w catalog years
  - click continue
- enter some minors w catalog years
  - click start planning (creates account w majors / minors and reroutes to `/planner`)

#### Test routing
- nav to different routes like `/planner`, `/`, and `/register` with different permutations of:
  - not logged in
  - logged in / no account created
  - logged in / account created

# TODO

- [x] Zustand setup
- [x] Implement Account Creation UI
- [x] Change Major/Minor `<Select>` -> `<Autocomplete>`
- [x] Implement auth routing (and data saving)
- [x] Write user server action tests
- [x] Write catalogYears server action tests
- [ ] ~~Better rendering performance / code clarity with [React Hook Form](https://react-hook-form.com/)?~~
- [ ] ~~Help (?) tooltips or modals~~
- [ ] ~~Curriculum select page~~

### CODE REVIEW

File: src/app/hooks/reactQuery.ts [hunk 1]
```diff
@@ -70,13 +130,41 @@ export function useProgramTypeOfYear(
...
   });
 }
+
+/**
+ * A React Query hook to fetch all catalog years, optionally for a specific
+ * program
+ * @param programName (Optional) Name of the program
+ * @returns React Query useQuery Hook for all catalog years
+ */
+export function useCatalogYears() {
```
Comment (fercevik729) [inline src/app/hooks/reactQuery.ts:146]: This is a little redundant having two hooks with very similar implementations

File: src/app/hooks/reactQuery.ts [hunk 2]
```diff
@@ -47,18 +67,58 @@ export function usePrograms() {
...
+ * @returns React Query useQuery Hook for all majors and minors
+ */
+export function useUnqiueProgramsInYear(catalogYear: string) {
+  const queryClient = useQueryClient();
+
+  return useQuery({
+    queryKey: ["uniquePrograms", catalogYear],
+    queryFn: async () => {
+      const res = await getPrograms(catalogYear);
+      return filterRedundantPrograms(res);
```
Comment (fercevik729) [inline src/app/hooks/reactQuery.ts:103]: Any reason we can't move the `filterRedundantPrograms` to the server?

File: src/app/register/(skipSetup)/page.tsx [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+      queryKey: ["programs"],
+      queryFn: () => getPrograms(),
+    });
+  }, [queryClient]);
+
+  return (
+    <>
+      <div>
+        <h2 className="mt-8 text-3xl font-bold leading-9 tracking-tight text-gray-900">
+          Start you&apos;re degree off right
```
Comment (fercevik729) [inline src/app/register/(skipSetup)/page.tsx:75]:
```suggestion
          Start your degree off right
```

File: src/app/register/(skipSetup)/page.tsx [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+              Pick from UCSC&apos;s suggested curriculums
+            </p>
+          </div>
+        </div>
+      </SelectBox>
+
+      <div className="h-5" />
+
+      <SelectBox
+        selected={skipSetup === true}
```
Comment (fercevik729) [inline src/app/register/(skipSetup)/page.tsx:110]:
```suggestion
        selected={skipSetup}
```

File: src/app/register/ProgramSelector.tsx [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+    });
+  };
+
+  // Delete a program from the list of selected programs
+  const handleDeleteProgram = (programId: number) => {
+    setError("");
+    deleteProgram(programId);
+  };
+
+  // NOTE: User thrown errors (more than one of same major) exist in addition to
```
Comment (fercevik729) [inline src/app/register/ProgramSelector.tsx:87]: If this is the case, we might want to use either React Hook Form or Material/Joy UI forms

File: src/lib/auth.ts [hunk 6]
```diff
@@ -5,55 +5,61 @@ import { NextAuthOptions } from "next-auth";
...
-            where: {
-              id: session.user.id,
-            },
-            select: {
-              defaultPlannerId: true,
-            },
-          })
-        )?.defaultPlannerId;
+
+        // QUESTION: Why is default planner id being set here? Should be in user
```
Comment (fercevik729) [inline src/lib/auth.ts:37]: It's helpful to have the default planner id in the session object to avoid having to refetch it again later on

File: src/app/register/ProgramSelector.tsx [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+
+  return (
+    <>
+      {/* Program, catalogYear, add button */}
+      <div className="flex flex-col md:flex-row gap-4">
+        <Autocomplete
+          value={minorValue}
+          options={programs}
+          getOptionLabel={(option) => option.name}
+          placeholder="Minor"
```
Comment (fercevik729) [inline src/app/register/ProgramSelector.tsx:105]: Placeholder should match program type

**Review by fercevik729 — CHANGES_REQUESTED** [review]
Minor issues with code redundancy and testing. Routing logic seems a little hard to follow at times and I'm not sure if the errors in the dev environment disappear in the production environment. We might need to bump to auth v5 even if it's still in beta and see if everything still works then go from there.

 Also since this PR lacks a curriculum/default planner selector I am hesitant to merge it into `main` for the time being. Merging into `dev` might be a better move here.

File: src/app/hooks/reactQuery.ts (hunk 2 above)
Comment (zacktraczyk) [reply src/app/hooks/reactQuery.ts:103]: We could definitely do that, would probably have to write a raw SQL query as prisma doesn't support the DISTINCT keyword in SQL. This way is nice as it saves database calls and derives the data from existing cache data if it exists, but does require the client to perform the work.

Up to you, if you think a dedicated server action with a SQL DISTINCT query is good going forward I can add it to this PR.

File: src/lib/auth.ts (hunk 6 above)
Comment (zacktraczyk) [reply src/lib/auth.ts:37]: Doesn't the default planner id change if you were to delete it? Mutable user data shouldn't be stored in the session, if you want it to always be available you can use Zustand store or react query if it is server state and set a stale time of infinity. 

Maybe I misunderstand the point of this value.

File: __tests__/actions/user.test.ts [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+        majors: true,
+      },
+    });
+
+    expect(userInDatabase).toBeDefined();
+    expect(userInDatabase?.id).toBe(userId);
+    expect(userInDatabase?.name).toBe("Sammy Slug");
+    expect(userInDatabase?.email).toBe("sammy@ucsc.edu");
+    expect(userInDatabase?.role).toBe("USER");
+    expect(userInDatabase?.majors).toHaveLength(0);
```
[the remark this answers is not in the document]
Comment (zacktraczyk) [reply __tests__/actions/user.test.ts:50]: I believe the `afterAll` function calls `prisma.user.deleteMany()` which deletes all the entries in the users table.

[the remark this answers is not in the document]
Comment (zacktraczyk) [reply __tests__/actions/user.test.ts:50]: I saw this convention in `major.test.ts` where basically the whole database is delete during the test suite run.

File: __tests__/actions/major.test.ts [hunk 9]
```diff
@@ -280,4 +282,72 @@ describe("Major actions", () => {
...
+    const catalogYearsObjs = await getCatalogYears();
+    const catalogYears = catalogYearsObjs.map((year) => year.catalogYear);
+
+    expect(catalogYears).toContain("2020-2021");
+    expect(catalogYears).toContain("2021-2022");
+    expect(catalogYears).toContain("2022-2023");
+    expect(catalogYears).toContain("2023-2024");
+  });
+
+  it("should return the available catalog years for a specific major", async () => {
```
[the remark this answers is not in the document]
Comment (zacktraczyk) [reply __tests__/actions/major.test.ts:329]: > I believe the `afterAll` function calls `prisma.user.deleteMany()` which deletes all the entries in the users table.

In a comment below I address this. I used the same conventions as already existing in the file where all database data is deleted after the test suite. I think there is a need for a test refactor that creates and deletes specific ids then clearing the whole database. Also may be nice to standardized the CRUD operations by creating uniform server actions.

File: src/app/register/ProgramSelector.tsx (hunk 5 above)
Comment (zacktraczyk) [reply src/app/register/ProgramSelector.tsx:87]: I tried using React Hook form but realized the use case doesn't make sense for this particular case. The way error / warning messages are displayed requires hooking into the error state so makes sense just to control it manually.

File: __tests__/actions/major.test.ts [hunk 10]
```diff
@@ -280,4 +282,72 @@ describe("Major actions", () => {
...
+      },
+    });
+
+    const catalogYearsObjs = await getCatalogYears();
+    const catalogYears = catalogYearsObjs.map((year) => year.catalogYear);
+
+    expect(catalogYears).toContain("2020-2021");
+    expect(catalogYears).toContain("2021-2022");
+    expect(catalogYears).toContain("2022-2023");
+    expect(catalogYears).toContain("2023-2024");
```
[the remark this answers is not in the document]
Comment (zacktraczyk) [reply __tests__/actions/major.test.ts:319]: see https://github.com/SlugPath/SlugPath/pull/338#discussion_r1543399061

File: __tests__/actions/major.test.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (fercevik729) [reply __tests__/actions/major.test.ts:322]: Actually this test suite already has an `afterEach` lifecycle hook that clears majors between tests so my mistake. Having standardized CRUD operations is a good idea, but I fear that it might lead to a lot of boilerplate-y code when we only need a subset of operations. Moreover, we don't have an API layer (yet) for which we need to standardize CRUD ops and can leverage the ad-hoc and customizable nature of server actions for now. We might need to do this for the `Permissions` table though because we were planning to create an API for that but that's kind of unrelated right now.

File: __tests__/actions/user.test.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (fercevik729) [reply __tests__/actions/user.test.ts:50]: You're right since there is only one test in this suite. If you do however add more tests to this suite consider using the `afterEach` hook to delete any intermediate data created in the test

File: src/app/register/(skipSetup)/page.tsx (hunk 4 above)
Comment (fercevik729) [reply src/app/register/(skipSetup)/page.tsx:110]: Up to you in that case

File: src/lib/auth.ts (hunk 6 above)
Comment (fercevik729) [reply src/lib/auth.ts:37]: No you're right, this is probably a cause of some subtle errors. I think it would be best to just use the `useUserDefaultPlannerId` hook to access it as it is technically server state

File: src/app/hooks/reactQuery.ts (hunk 2 above)
Comment (fercevik729) [reply src/app/hooks/reactQuery.ts:103]: If there is no need for duplicate programs returned by the `getPrograms` server action, we can leverage the distinct field in the parameter object of `findMany` and avoid having to use `filterRedundantPrograms` altogether as follows:
```ts
const distinctCities = await prisma.user.findMany({
  select: {
    city: true,
    country: true,
  },
  distinct: ['city'],
}) 
```

File: __tests__/actions/major.test.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (zacktraczyk) [reply __tests__/actions/major.test.ts:322]: sounds good

File: src/app/hooks/reactQuery.ts (hunk 2 above)
Comment (zacktraczyk) [reply src/app/hooks/reactQuery.ts:103]: Prisma does have a distinct option but it doesn't actually use SQL Distinct but rather in-memory sorting after the query which is not as well optimized ([prisma docs explains it here](https://www.prisma.io/docs/orm/prisma-client/queries/aggregation-grouping-summarizing#distinct-under-the-hood)). Not a super big performance hit so I guess it doesn't matter to much but may be beneficial to avoid.

Also there is a use case for both all programs and distinct programs. In my `ProgramSelector.tsx` I reduced all the db calls to just one action and derived the necessary data so unique would not work here. I'll add a server action to my PR for the permissions modal that uses distinct programs, but maybe its a better tradeoff to move redundancy filtering to the client and utilize the react query cache, not sure.

``` typescript
/**
 * A React Query hook to fetch all unique majors and minors
 * @returns React Query useQuery Hook for all majors and minors
 */
export function useUniquePrograms() {
  const queryClient = useQueryClient();

  return useQuery({
    queryKey: ["uniquePrograms"],
    queryFn: async () => {
      const res = await getPrograms();
      return filterRedundantPrograms(res);
    },
    // Use cache to avoid refetching data
    initialData: () => {
      const programs: Program[] | undefined = queryClient.getQueryData([
        "programs",
      ]);
      return programs ? filterRedundantPrograms(programs) : [];
    },
    placeholderData: [],
    staleTime: Infinity, // Programs are static
    refetchOnMount: false,
  });
}
```

File: src/app/hooks/reactQuery.ts (hunk 1 above)
Comment (zacktraczyk) [reply src/app/hooks/reactQuery.ts:144]: which part exactly is redundant, `useCatalogYears` and `useCatalogYearsForProgram`? I was thinking it was clearer to have different functions for some reason but I think you're right it makes sense to combine and have `programName` as an optional.

Comment (zacktraczyk) [reply src/app/hooks/reactQuery.ts:144]: I used this pattern for a bunch of hooks I can undo them all

Comment (fercevik729) [reply src/app/hooks/reactQuery.ts:144]: Might be better to merge all "redundant" hooks especially if they use the same underlying server actions

**Review by fercevik729 — APPROVED** [review]
